### ISSUE TITLE
Add dynamite chaining

### ISSUE THREAD

**Issue by rmarquis** [etlegacy/etlegacy#1487]
When many dynamites are set in a single location, dyna chaining blows up all dynamite that are in the blast radius of the first exploding dynamite.

There are a couple of reason to implement this:
* this prevents teammates to be blown up by the second or third dynamite when the first has done its job (f.e., exploding the siwa oasis wall).
* this prevents the "dyna exploit" where players would put two dynamites to blow up a multi-stage objective in a single move, bypassing the intent of the mapper (note: this is still possible if extra dyna are set outside of the first blast radius)
* this greatly simplifies the implementation of the dynamite indicator #1242.

**Comment by realkemon** [comment]
On Discord there was a discussion about pros and cons of this.

Simply detonating all active dynamites in the trigger area changes gameplay behaviour for several entities. Especially, func_constructibles with multiple implemented stages (desstages and constages). These are designed by the game to require multiple dynamites. Removing active dynamites upon changing the first stage means that the second stage remains active. This is in that regard a gameplay change, that now the dynamiting team needs to defend the area for a lot longer time. Which could mean surviving an additional entire enemy spawnwave while they themselves are further away from their spawn. Previously you only needed to defend ticking dynamites. Now you need to defend the ticking dynamites and the engineer to then plant another dynamite which you need to defend.

There are two possible approaches for this.
One would be to make an exception for entity setups purposely designed to be multi-staged like the Fueldump bridge or depot defenses. These entities are `func_constructibles` with multiple constages and desstages implemented.

Another approach would be to only implement dynamite chain for func_explosives. These entities are by design only one time things (e.g. Old City Wall in Oasis, Main/Side Entrance in Radar). Func_constructibles on the other hand can be rebuilt. Another active dynamite ticking means that the defending team cannot simply rebuild the defense that easily. Either they wait for the dynamite to blow up which means they have to defend the area manually, just like the attacker have to when their dynamite is planted. Or they move out of cover to defuse that additional dynamite which means putting the engineer in harms way.

Most of the time defenses like that are easily rebuilt as it is, because the materials are on the defender side of the barrier so removing all active dynamites plays into the hands of the defense.

TL;DR: 
#1 Make an exception to the dynamite chain for func_constructibles with multiple stages implemented (desstages, constages)
#2 Only apply dynamite chain for func_explosive which by design are a single use entity

**Comment by Helldembez** [comment]
After arming a dynamite you drop a non armed dynamite nearby around the time the first dyna explodes it deletes the non armed dynamite with it. I will add a demo with timestamps later today.

**Comment by rmarquis** [comment]
How is that an issue, if I may ask? The objective would be destroyed, so it doesn't matter if the non armed dyna is gone.

**Comment by Helldembez** [comment]
Sorry, I did not mention that it was a misplant.

**Comment by rmarquis** [comment]
Ah, you mean the exploding dyna was not on objective in the first place? There is a check to ensure dyna are on the same objective in the dyno chaining system, but it only accounts for armed dyna.

**Comment by Helldembez** [comment]
[2021-08-03-223713-frostbite.zip](https://github.com/etlegacy/etlegacy/files/6947567/2021-08-03-223713-frostbite.zip)

Here you should see exactly what happened and how.

**Comment by rmarquis** [comment]
From @Aciz 

> There is absolutely no check for it anywhere, and it's set to explode the dynamites 250ms after the first one, so it's making the original "exploit" of planting multiple dynamites on objective even worse.

> I started poking around the code and was wondering what's the intended behavior of dynamite chaining on multi-stage constructibles, should it
>
> * not explode dynamites until we're at the last stage? So essentially disable dynamite chaining until the next explosion would completely destroy it
> or
> * remove the dynamites from the objective without exploding them? This would be a gameplay change from vanilla with no way to disable it, keeping in mind what @realkemon wrote in the [above comment](https://github.com/etlegacy/etlegacy/issues/1487#issuecomment-656621909).

**Comment by rmarquis** [comment]
I kinda see the multi dyna as an exploit, but after giving some thought I agree with you that the proposed solution might be the least invasive. In any case it's much better than the current buggy behaviour which makes that exploit even worst.

There is one aspect I'd like to discuss further and that is the optional cvar to remove dyna instead of blowing them up. Since the cvar doesn't disable the behaviour to give a vanilla behaviour (unlike what I initially thought), it kinda feels over-engineered to me. In essence, the difference between both behaviour is the amount of damage given (1 dyna, vs all dyna), and to me there is very little difference. I think we should apply one of the behaviour and drop the cvar entirely.

Option 1 is blowing all dyna (increased damage output), while option 2 is removing dyna (only damage from 1 dyna). The first option might feel more realistic, but the second one might have less side effects on gameplay. F.e., I can imagine a engi with flak jacket surviving a single dyna blast, but he might not in the case of multiple dyna blast.

Would there be some opposition to dropping this cvar, or some clear advantage I'm not seeing? The option of simply removing dyna by default seems fine to me. This doesn't concern the behaviour on multi-stage discussed above, and here I'm more concerned of standardizing gameplay as much possible across servers.

**Comment by Aciz** [comment]
Isn't that exactly what `g_dynamiteChaining 1` does? It explodes the main dynamite, then removes the others without them dealing damage.

**Comment by rmarquis** [comment]
Yes. I am suggesting to remove `g_dynamiteChaining` and make one of its behavior as default. I might have a slight preference for `1`, but `0` would be also fine.

**Comment by Aciz** [comment]
I see.  I don't really have any strong opinion on that, but if either behavior is to be forced, I do lean on `1` too. Exploding can be especially problematic on crowded public servers such as TM main, where people might plant like 5 dynamites on an objective. Forcing `0` would also probably not be wanted for competitive play (though it's very rare to plant multiple dynamites at the same time because of charge restrictions and usually only having one engineer who doesn't use his charge on rifle nades).

**Comment by rmarquis** [comment]
Removed the cvar, while keeping its behaviour.

Dynamite chaining now remove the drawback of not having it (blowing up teammates after the obj was dynamited) without side effects on multi-stage objectives. Closing the ticket now.
